django-verdant/rca/management/commands/archive_events.py

```diff
@@ -0,0 +1,161 @@
+        ),
+        make_option(
+            '--permanent',
+            action='store_true',
+            dest='permanent',
+            default=True,
+            help="Create permanent redirects (301)."
+        ),
+    )
+
+    def create_redirect(self, e, target_page, permanent=True):
+        new_redirect = Redirect()
+        new_redirect.site = e.get_site()
+        # Path includes 'home' and a trailing slash
+        # both must be removed.
+        bare_path = '/' + '/'.join(e.url_path.split('/')[2:][:-1])
+        new_redirect.old_path = bare_path
+        new_redirect.is_permanent = permanent
+        new_redirect.redirect_page = target_page
+        new_redirect.save()
+        self.event_redirects[e.pk] = new_redirect.id
+
+    def get_csv_rows(self, events):
+        rows = []
+        headers = [
+            'Event id',
+            'Event title',
+            'Event URL',
+            'Dates from',
+            'Dates to',
+            'Redirect',
+        ]
+        rows.append(headers)
+        for e in events:
+            row = []
+            row.append(e.pk)
+            row.append(e.title.encode('utf8'))
+            row.append(e.url)
+            dates = e.dates_times.order_by('date_from').values_list('date_from', 'date_to')
+            date_from = []
+            for d in dates:
+                try:
+                    date_from.append(d[0].strftime("%d/%m/%Y"))
+                except Exception:
+                    date_from = ''
+
+            date_to = []
+            for d in dates:
+                try:
+                    date_to.append(d[1].strftime("%d/%m/%Y"))
+                except Exception:
+                    date_to = ''
+
+            row.append(', '.join(date_from))
+            row.append(', '.join(date_to))
+            if self.event_redirects:
+                row.append('{}/admin/redirects/{}/'.format(
+                    e.get_site().root_url,
+                    self.event_redirects[e.pk])
+                )
+            else:
+                row.append('')
+            rows.append(row)
+
+        return rows
+
+    def handle(self, **options):
+        verbose = options['verbose']
+        examples = options['examples']
+        output_csv = options['output_csv']
+        permanent = options['permanent']
+        dry_run = options['dry_run']
+
+        event_index = EventIndex.objects.first()
+
+        if csv:
+            verbose = False
+            examples = False
+
+        events = EventItem.objects.live()
+        events_count = events.count()
+        two_years_ago = date.today() - relativedelta(years=2)
+
+        events = events.exclude(dates_times__date_to__gt=two_years_ago)
+        events = events.exclude(dates_times__date_from__gt=two_years_ago)
+
+        if not dry_run:
+            with transaction.atomic():
+                [e.unpublish() for e in events]
+            [self.create_redirect(e, event_index, permanent) for e in events]
```

Comment:
yeah, just love those comprehensions :)

I had assumed that if unpublishing an event had caused an error, it wouldn't get to the create_redirect call, is that incorrect? Either way, it would seem to make sense to put it in the atomic bit.

Reply:
If the error caused the script to halt there then I don't think it'd go further, but if it was swallowed in some way then it could end up running the redirects line.